### DISCUSSION TITLE
Force pos tag in tagger and morphologizer

### DISCUSSION THREAD

**Discussion by pdimiel** [explosion/spaCy#8260] [Help: Coding & Implementations] [answered]
Hello everyone.

I am trying to add a custom partial tagger module in the pipeline.
The goal is to be able to force some pos tags according to rules or the output of an external pos tagger but not for all the tokens.

In spacy 3.0.6 I have the following as an example.
The actual function will use complex rules or given tags from another resource.

```
@Language.component("custom_tagger")
def custom_tagger(doc):
    for i, token in enumerate(doc):
        if token.orth_.lower() == "είδα":
            doc[i].tag_ = "VERB"
        elif token.orth_.lower() == "saw":
            doc[i].tag_ = "NN"

    return doc
```

I have a test function:

```
def process_and_print(given_nlp, text):
    print(nlp.pipe_names)
    doc = given_nlp(text)
    for sent in doc.sents:
        print(sent.text)
        for token in sent:
            print(token, token.tag_)
    print()
```

and I use it for two languages, English (that use a tagger) and Greek (that use a morphologizer) as:

```
sample = "I saw you cutting the tree."
nlp = spacy.load("en_core_web_sm")
process_and_print(nlp, sample)
nlp.add_pipe("custom_tagger", before="tagger")  # Insert before the tagger
process_and_print(nlp, sample)

sample = "Είδα ένα δέντρο."
nlp = spacy.load("el_core_news_sm")
process_and_print(nlp, sample)
nlp.add_pipe("custom_tagger", before="morphologizer")  # Insert before the morphologizer
process_and_print(nlp, sample)
```

In the English case, the output is the expected

> ['tok2vec', 'tagger', 'parser', 'ner', 'attribute_ruler', 'lemmatizer']
> I saw you cutting the tree.
> I PRP
> **saw VBD**
> you PRP
> cutting VBG
> the DT
> tree NN
> . .
> 
> ['tok2vec', 'custom_tagger', 'tagger', 'parser', 'ner', 'attribute_ruler', 'lemmatizer']
> I saw you cutting the tree.
> I PRP
> **saw NN**
> you PRP
> cutting VBG
> the DT
> tree NN
> . .

In the Greek case, there is no effect.

> ['tok2vec', 'morphologizer', 'parser', 'ner', 'attribute_ruler', 'lemmatizer']
> Είδα ένα δέντρο.
> **Είδα NOUN**
> ένα DET
> δέντρο NOUN
> . PUNCT
> 
> ['tok2vec', 'custom_tagger', 'morphologizer', 'parser', 'ner', 'attribute_ruler', 'lemmatizer']
> Είδα ένα δέντρο.
> **Είδα NOUN**
> ένα DET
> δέντρο NOUN
> . PUNCT

I suppose that it has something to do with the use of morphologizer instead of a tagger.

If that is the issue, is there a way to feed (force) coarse grained tags to the morphologizer and let it decide the fine grained details?

Thanks in advance

**Comment by polm** [comment]
This seems to be an interaction with the `attribute_ruler`, not the morphologizer. 

This gives what you are getting now:

    nlp.add_pipe("custom_tagger", before="attribute_ruler")

But changing `before` to `after` gives the result you want. 

> If that is the issue, is there a way to feed (force) coarse grained tags to the morphologizer and let it decide the fine grained details?

I don't think it can do that. I'm not really familiar with the Greek pipeline, but it doesn't have fine grained tags for some reason. Instead the attribute ruler sets the tag based on the pos, where the pos is assigned by the morphologizer. So as a workaround you can assign pos instead of tag, or you can put your component after the attribute ruler. 

Do note that your component seems like a special case of an attribute ruler, so you may want to check [the docs for that](https://spacy.io/api/attributeruler).

**Comment by adrianeboyd** [comment, marked as answer]
I think this is due to an inconsistency between the tagger and morphologizer: the tagger doesn't overwrite any existing tags but the morphologizer does overwrite everything.

Right now, your best option is to add a custom component (or rules in the attribute ruler) **after** the morphologizer as a post-correction step.

The predictions that the tagger and morphologizer make are not influenced by any existing annotation, so the final annotation is the same if you move this from a pre-processing step to a post-processing step.

In the future, we'd like to make this more consistent and configurable across components, as with the `overwrite` option in the entity ruler and lemmatizer.

**Comment by pdimiel** [comment]
Indeed, moving the custom component after the morphologizer does alter the coarse grained tag but leaves unchanged the fine-grained tag (found in the morph attribute). Thus the final morph information is inconsistent with the pos tag.

I was hoping to pass coarse grained tags through rules and/or other resources and leave the morphologizer decide the actual morph attribute, i.e force to treat as Verb and let it decide tense, voice, mood etc.

For the time being I cannot implement a full morphologizer to override the existing one and I need the full morphological info as is is important for the Greek language processing.

I fully understand the "inconsistency" you mention since the morphologizer is a new module in the pipeline and I am looking forward to the adding of new features and configurations.

**Reply by adrianeboyd** [reply]
The morphologizer component is just not that sophisticated. Underneath it's a pretty simple tagger that predicts POS+MORPH together as one combined tag for each token, and it isn't able to take any partial information into account, like predicting MORPH based on partial POS.

If you have custom rules for POS, it would make sense to provide both POS and MORPH if possible, so those cases can stay consistent.
